pkg/services/pluginsintegration/cachekvstore/namespacedstore.go

```diff
@@ -0,0 +1,141 @@
+package cachekvstore
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/kvstore"
+)
+
+// keyLastUpdated is the key used to store the last updated time.
+const keyLastUpdated = "last_updated"
+
+// NamespacedStore is a Store that stores data in a *kvstore.NamespacedKVStore.
+// It uses the provided StoreKeyGetter to determine the underlying store's key to use for a given key.
+// It also stores a last updated time, which is unique for all the keys and is updated on each call to `Set`,
+// and can be used to determine if the data is stale.
+type NamespacedStore struct {
+	// kv is the underlying KV store.
+	kv *kvstore.NamespacedKVStore
+
+	// storeKeyGetter is a function that returns the underlying store's key to use for a given key.
+	// This allows to modify the key for the underlying storage.
+	// This is only used for actual data, not for the last updated time key.
+	storeKeyGetter StoreKeyGetter
+}
+
+// DefaultStoreKeyGetter is the default StoreKeyGetterFunc, which returns the key as-is.
+var DefaultStoreKeyGetter = StoreKeyGetterFunc(func(k string) string {
+	return k
+})
+
+// PrefixStoreKeyGetter returns a StoreKeyGetterFunc that returns the key with a prefix.
+// It is used to prefix all keys in the underlying store with a fixed prefix.
+func PrefixStoreKeyGetter(prefix string) StoreKeyGetterFunc {
+	return func(k string) string {
+		return prefix + k
+	}
+}
+
+// NamespacedStoreOpt is an option for NewNamespacedStore that modifies the store passed to it.
+type NamespacedStoreOpt func(store *NamespacedStore)
+
+// WithStoreKeyGetter sets the StoreKeyGetter to use.
+func WithStoreKeyGetter(g StoreKeyGetter) NamespacedStoreOpt {
+	return func(store *NamespacedStore) {
+		store.storeKeyGetter = g
+	}
+}
```

Comment:
are all of this necessary to be exported? To simplify it you can just replace `NamespacedStoreOpt` with a `prefix string` and just expose:

```
// NewNamespacedStore creates a new NamespacedStore using the provided underlying KVStore, namespace and optionally a prefix.
func NewNamespacedStore(kv kvstore.KVStore, namespace string, prefix string) plugins.KeyStore {
```

Honestly I think that the prefix is not needed but we can keep it to avoid changes in the database.